File: verifier-cli/src/main/java/com/jetbrains/pluginverifier/verifiers/util/ParentsVisitor.java
package com.jetbrains.pluginverifier.verifiers.util;

import com.google.common.base.Predicate;
import com.intellij.structure.resolvers.Resolver;
import com.jetbrains.pluginverifier.error.VerificationError;
import org.jetbrains.annotations.NotNull;
import org.objectweb.asm.tree.ClassNode;

import java.util.*;


/**
 * @author Sergey Patrikeev
 */
class ParentsVisitor {

  private final Map<String, Set<String>> myParentsCache = new HashMap<String, Set<String>>();

  private final Resolver myResolver;

  ParentsVisitor(@NotNull Resolver resolver) {
    myResolver = resolver;
  }

  /**
   * Iterates through the inheritance tree of the specified class and collects
   * all the unresolved parent-classes
   *
   * @param className         start class
   * @param excludedPredicate whether to treat given class as excluded (and not process it at all)
   * @return set of unresolved classes
   * @throws IllegalArgumentException in case starting class is not found in the {@link #myResolver resolver}
   */
  @NotNull
  Set<String> collectUnresolvedParents(@NotNull String className,
                                       @NotNull Predicate<String> excludedPredicate) throws IllegalArgumentException, VerificationError {
    ClassNode classNode = VerifierUtil.findClass(myResolver, className);
    if (classNode == null) {
      throw new IllegalArgumentException(className + " should be found in the resolver " + myResolver);
    }

    if (myParentsCache.containsKey(className)) {
      return myParentsCache.get(className);
    }

    Set<String> allParents = new HashSet<String>();

    String superName = classNode.superName;
    if (superName != null) {
      if (!excludedPredicate.apply(superName)) {
        if (VerifierUtil.findClass(myResolver, superName) != null) {
          allParents.addAll(collectUnresolvedParents(superName, excludedPredicate));
        } else {
          allParents.add(superName);
        }
      }
    }

    @SuppressWarnings("unchecked")
    List<String> interfaces = classNode.interfaces;
    if (interfaces != null) {
      for (String anInterface : interfaces) {
        if (!excludedPredicate.apply(anInterface)) {
          if (VerifierUtil.findClass(myResolver, anInterface) != null) {
            allParents.addAll(collectUnresolvedParents(anInterface, excludedPredicate));
          } else {
            allParents.add(anInterface);
          }
        }
      }
    }

    myParentsCache.put(className, allParents);
    return allParents;
  }

}

File: verifier-core/src/main/java/com/jetbrains/pluginverifier/utils/ParentsVisitor.kt
package com.jetbrains.pluginverifier.utils

import com.intellij.structure.resolvers.Resolver
import com.jetbrains.pluginverifier.api.VContext

import java.util.*
import java.util.function.Predicate


/**
 * @author Sergey Patrikeev
 */
internal class ParentsVisitor(private val myResolver: Resolver, private val myCtx: VContext) {

  private val myParentsCache = HashMap<String, Set<String>>()

  /**
   * Iterates through the inheritance tree of the specified class and collects
   * all the unresolved parent-classes

   * @param className         start class
   * *
   * @param excludedPredicate whether to treat given class as excluded (and not process it at all)
   * *
   * @return set of unresolved classes
   * *
   * @throws IllegalArgumentException in case starting class is not found in the [resolver][.myResolver]
   */
  @Throws(IllegalArgumentException::class)
  fun collectUnresolvedParents(className: String,
                               excludedPredicate: Predicate<String>): Set<String> {
    val classNode = VerifierUtil.findClass(myResolver, className, myCtx) ?: throw IllegalArgumentException(className + " should be found in the resolver " + myResolver)

    if (myParentsCache.containsKey(className)) {
      return myParentsCache[className]!!
    }

    val allParents = HashSet<String>()

    val superName = classNode.superName
    if (superName != null) {
      if (!excludedPredicate.test(superName)) {
        if (VerifierUtil.findClass(myResolver, superName, myCtx) != null) {
          allParents.addAll(collectUnresolvedParents(superName, excludedPredicate))
        } else {
          allParents.add(superName)
        }
      }
    }

    val interfaces = classNode.interfaces
    if (interfaces != null) {
      @Suppress("UNCHECKED_CAST")
      for (anInterface in interfaces as List<String>) {
        if (!excludedPredicate.test(anInterface)) {
          if (VerifierUtil.findClass(myResolver, anInterface, myCtx) != null) {
            allParents.addAll(collectUnresolvedParents(anInterface, excludedPredicate))
          } else {
            allParents.add(anInterface)
          }
        }
      }
    }

    myParentsCache.put(className, allParents)
    return allParents
  }

}
